be/src/storage/persistent_index.cpp

```diff
@@ -1216,6 +1254,56 @@ Status PersistentIndex::_flush_l0() {
     return writer.finish();
 }
 
+Status PersistentIndex::_rebuild() {
+    _offset = 0;
+    _page_size = 0;
+    auto l0_st = MutableIndex::create(_key_size);
+    if (!l0_st.ok()) {
+        return l0_st.status();
+    }
+    _l0 = std::move(l0_st).value();
+
+    auto idx_file_path = strings::Substitute("$0/index.l1.$1.$2", _path, _version.major(), _version.minor());
+    fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+    std::unique_ptr<fs::ReadableBlock> rblock;
+    RETURN_IF_ERROR(block_mgr->open_block(idx_file_path, &rblock));
+    auto l1_st = ImmutableIndex::load(std::move(rblock));
+    if (!l1_st.ok()) {
+        return l1_st.status();
+    }
+    _l1 = std::move(l1_st).value();
+    return Status::OK();
+}
+
+// check _l0 should be flush or not, if not, return
+// if _l0 should be flush, there are two conditions:
+//   1. _l1 is not exist, _flush_l0 and build _l1
+//   2. _l1 is exist, merge _l0 and _l1
+// rebuild _l0 and _l1
+// In addition, there may be io waste because we write wal(dump snapshot) first and
+// do _flush_l0 or merge compaction.
+Status PersistentIndex::_check_and_flush_l0() {
+    size_t kv_size = _key_size + sizeof(IndexValue);
+    size_t l0_mem_size = kv_size * _l0->size();
+    uint64_t l1_file_size = 0;
+    if (_l1 != nullptr) {
+        _l1->file_size(&l1_file_size);
+    }
+    if (l0_mem_size <= l0_flush_size_min &&
+        ((l0_mem_size <= l0_snapshot_size_max) || (l1_file_size / l0_mem_size > 10))) {
+        return Status::OK();
+    }
+    // flush _l0
+    if (_l1 == nullptr) {
+        RETURN_IF_ERROR(_flush_l0());
+    } else {
+        RETURN_IF_ERROR(_merge_compaction());
+    }
+    // rebuild _l0 and _l1
+    RETURN_IF_ERROR(_rebuild());
```

Comment:
I think the more clear order is:
commit:

1. check condition, chose which action to do: snaphot/wal/flush/merge_compaction
2. update meta according to different action

on_commit:

if necessary, e.g. flush/merge_compaction occurred
    * reload(_rebuild) from new versions meta&file
    * clear old files

Reply:
> I think the more clear order is: commit:
> 
> 1. check condition, chose which action to do: snaphot/wal/flush/merge_compaction
> 2. update meta according to different action
> 
> on_commit:
> 
> if necessary, e.g. flush/merge_compaction occurred * reload(_rebuild) from new versions meta&file * clear old files

done, change in https://github.com/StarRocks/starrocks/pull/4172/commits/19bb730532d645dd8192ac70522ad282327f364c